client/CServerHandler.cpp

```diff
@@ -245,88 +256,97 @@ void CServerHandler::startLocalServerAndConnect()
 
 	th->update(); //put breakpoint here to attach to server before it does something stupid
 
-	justConnectToServer(localhostAddress, 0);
+	connectToServer(getLocalHostname(), getLocalPort());
 
 	logNetwork->trace("\tConnecting to the server: %d ms", th->getDiff());
 }
 
-void CServerHandler::justConnectToServer(const std::string & addr, const ui16 port)
+void CServerHandler::connectToServer(const std::string & addr, const ui16 port)
 {
-	state = EClientState::CONNECTING;
-	while(!c && state != EClientState::CONNECTION_CANCELLED)
-	{
-		try
-		{
-			logNetwork->info("Establishing connection...");
-			c = std::make_shared<CConnection>(
-					addr.size() ? addr : getHostAddress(),
-					port ? port : getHostPort(),
-					NAME, uuid);
-
-			nextClient = std::make_unique<CClient>();
-			c->iser.cb = nextClient.get();
-		}
-		catch(std::runtime_error & error)
-		{
-			logNetwork->warn("\nCannot establish connection. %s Retrying in 1 second", error.what());
-			boost::this_thread::sleep_for(boost::chrono::milliseconds(1000));
-		}
-	}
+	logNetwork->info("Establishing connection to %s:%d...", addr, port);
+	setState(EClientState::CONNECTING);
+	serverHostname = addr;
+	serverPort = port;
 
-	if(state == EClientState::CONNECTION_CANCELLED)
+	if (!isServerLocal())
 	{
-		logNetwork->info("Connection aborted by player!");
-		return;
+		Settings remoteAddress = settings.write["server"]["remoteHostname"];
+		remoteAddress->String() = addr;
+
+		Settings remotePort = settings.write["server"]["remotePort"];
+		remotePort->Integer() = port;
 	}
 
-	c->handler = std::make_shared<boost::thread>(&CServerHandler::threadHandleConnection, this);
+	networkHandler->connectToRemote(*this, addr, port);
+}
+
+void CServerHandler::onConnectionFailed(const std::string & errorMessage)
+{
+	assert(getState() == EClientState::CONNECTING);
+	boost::mutex::scoped_lock interfaceLock(GH.interfaceMutex);
 
-	if(!addr.empty() && addr != getHostAddress())
+	if (isServerLocal())
 	{
-		Settings serverAddress = settings.write["server"]["server"];
-		serverAddress->String() = addr;
+		// retry - local server might be still starting up
+		logNetwork->debug("\nCannot establish connection. %s. Retrying...", errorMessage);
+		networkHandler->createTimer(*this, std::chrono::milliseconds(100));
```

Comment:
Yes and no.
Yes, effectively fixed in #3609. When server runs in 'thread' mode, first attempt will always succeed, because server is guaranteed to listen for connection when we reach this code.

But for 'process' mode - no. I mean, boost::process might provide such functionality but needs research & testing. Considering that 'process' mode is kind of deprecated don't see need to spend time on it to be honest.